src/mocks/ScrollChainMockFinalize.sol

```diff
@@ -27,95 +27,52 @@ contract ScrollChainMockFinalize is ScrollChain {
-        withdrawRoots[_batchIndex] = _withdrawRoot;
+        // verify blob versioned hash
+        bytes32 _blobVersionedHash = BatchHeaderV1Codec.getBlobVersionedHash(batchPtr);
+        _checkBlobVersionedHash(_blobVersionedHash, _blobDataProof);
 
         // Pop finalized and non-skipped message from L1MessageQueue.
-        _popL1Messages(
-            BatchHeaderV0Codec.getSkippedBitmapPtr(memPtr),
-            BatchHeaderV0Codec.getTotalL1MessagePopped(memPtr),
-            BatchHeaderV0Codec.getL1MessagePopped(memPtr)
+        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
+        _popL1MessagesMemory(
+            BatchHeaderV1Codec.getSkippedBitmapPtr(batchPtr),
+            _totalL1MessagesPoppedOverall,
+            BatchHeaderV0Codec.getL1MessagePopped(batchPtr)
         );
 
-        emit FinalizeBatch(_batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
+        _afterFinalizeBatch(_totalL1MessagesPoppedOverall, _batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
     }
 
-    /// @notice Finalize 4844 batch without proof, See the comments of {ScrollChain-finalizeBatchWithProof4844}.
-    function finalizeBatch4844(
+    /// @notice Finalize bundle without proof, See the comments of {ScrollChain-finalizeBundleWithProof}.
+    function finalizeBundle(
         bytes calldata _batchHeader,
-        bytes32 _prevStateRoot,
         bytes32 _postStateRoot,
-        bytes32 _withdrawRoot,
-        bytes calldata _blobDataProof
+        bytes32 _withdrawRoot
     ) external OnlyProver whenNotPaused {
-        if (_prevStateRoot == bytes32(0)) revert ErrorPreviousStateRootIsZero();
         if (_postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
 
-        // compute batch hash and verify
-        (uint256 memPtr, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
-        bytes32 _blobVersionedHash = BatchHeaderV1Codec.getBlobVersionedHash(memPtr);
-
-        // Calls the point evaluation precompile and verifies the output
-        {
-            (bool success, bytes memory data) = POINT_EVALUATION_PRECOMPILE_ADDR.staticcall(
-                abi.encodePacked(_blobVersionedHash, _blobDataProof)
-            );
-            // We verify that the point evaluation precompile call was successful by testing the latter 32 bytes of the
-            // response is equal to BLS_MODULUS as defined in https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile
-            if (!success) revert ErrorCallPointEvaluationPrecompileFailed();
-            (, uint256 result) = abi.decode(data, (uint256, uint256));
-            if (result != BLS_MODULUS) revert ErrorUnexpectedPointEvaluationPrecompileOutput();
-        }
+        // compute pending batch hash and verify
+        (, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
+        if (_batchIndex <= lastFinalizedBatchIndex) revert ErrorBatchIsAlreadyVerified();
```

Comment:
We should also check if the last batch is committed.

Because during adding unit tests, when the data are ready in db, and rollup-relayer calls `finalizeBundle` before committing these batches, the finalized height in the contract would be directly updated to `_batchIndex`, thus some batches cannot be committed to the contract.

Reply:
It is also checked in `_loadBatchHeader`.

Reply:
got it. after revisiting, I found the real issue is that if rollup-relayer calls finalizeBundle (without verifier), it will finalize <= batch index batches (including unfinalized codecv2 batches), if some codecv2 batches are not finalized yet, rollup-relayer will keep complaining that it cannot finalize these batches.